core/src/main/scala/kafka/server/epoch/LeaderEpochFileCache.scala

```diff
@@ -81,33 +81,44 @@ class LeaderEpochFileCache(topicPartition: TopicPartition, leo: () => LogOffsetM
   }
 
   /**
-    * Returns the End Offset for a requested Leader Epoch.
+    * Returns the Leader Epoch and the End Offset for a requested Leader Epoch.
     *
-    * This is defined as the start offset of the first Leader Epoch larger than the
-    * Leader Epoch requested, or else the Log End Offset if the latest epoch was requested.
+    * The Leader Epoch returned is the largest epoch less than or equal to the requested Leader
+    * Epoch. The End Offset is the end offset of this epoch, which is defined as the start offset
+    * of the first Leader Epoch larger than the Leader Epoch requested, or else the Log End
+    * Offset if the latest epoch was requested.
     *
     * During the upgrade phase, where there are existing messages may not have a leader epoch,
     * if requestedEpoch is < the first epoch cached, UNSUPPORTED_EPOCH_OFFSET will be returned
     * so that the follower falls back to High Water Mark.
     *
-    * @param requestedEpoch
-    * @return offset
+    * @param requestedEpoch requested leader epoch
+    * @return leader epoch and offset
     */
-  override def endOffsetFor(requestedEpoch: Int): Long = {
+  override def endOffsetFor(requestedEpoch: Int): (Int, Long) = {
     inReadLock(lock) {
-      val offset =
-        if (requestedEpoch == latestEpoch) {
-          leo().messageOffset
-        }
-        else {
-          val subsequentEpochs = epochs.filter(e => e.epoch > requestedEpoch)
+      val epochAndOffset =
+        if (requestedEpoch == UNDEFINED_EPOCH) {
+          // this is ok in case of bootstrapping follower sending a request with undefined epoch
+          // since it does not have any epochs; but otherwise not expected. So, we log as info
+          info(s"Requested end offset for undefined leader epoch for partition $topicPartition. " +
```

Comment:
Do we need to log this?

Reply:
You are right, probably not here. I was thinking to log the case where the follower somehow loses the leader epoch sequence file or something that causes it sending the request with undefined epoch, which could potentially cause log divergence. But now I thinking we should log error on the follower if we detect such problem.